src/compile/projection/assemble.ts

```diff
@@ -0,0 +1,56 @@
+import {error} from 'util';
+import {contains} from '../../util';
+import {isVgSignalRef, VgProjection, VgSignalRef} from '../../vega.schema';
+import {FacetModel} from '../facet';
+import {isConcatModel, isLayerModel, isRepeatModel, isUnitModel, Model, ModelWithField} from '../model';
+import {UnitModel} from '../unit';
+
+export function assembleProjections(model: Model): VgProjection[] {
+  if (isLayerModel(model) || isConcatModel(model) || isRepeatModel(model)) {
+    return assembleProjectionsForModelAndChildren(model);
+  } else {
+    return assembleProjectionForModel(model);
+  }
+}
+
+export function assembleProjectionsForModelAndChildren(model: Model): VgProjection[] {
+  return model.children.reduce((projections, child) => {
+    return projections.concat(child.assembleProjections());
+  }, assembleProjectionForModel(model));
+}
+
+export function assembleProjectionForModel(model: Model): VgProjection[] {
+  const component = model.component.projection;
+  if (!component || component.merged) {
+    return [];
+  }
+
+  const projection = component.combine();
+  const {name, ...rest} = projection;
```

Comment:
not sure why you have to extract `name` and `rest` here, simply to merge them again at the end?

Reply:
I think I do for some weird reason to do with types. if I remove them from the bottom and just extend `projection` with the return, the type check yells at me. Basically a vega projection is all the properties of the vega-lite projection + name, size, and fit (which are algorithmically decided)

Reply:
This is fine. I added a comment.